crypto/lms/lms_verify.c

```diff
@@ -0,0 +1,168 @@
+         *   Tc(parent) = H(I || node_q || 0x8383 || paths[i][n] || Tc(right) OR
+         *   Tc(parent) = H(I || node_q || 0x8383 || Tc(left) || paths[i][n])
+         */
+        if (!EVP_MD_CTX_copy_ex(ctx, ctxI)
+                || !EVP_DigestUpdate(ctx, qbuf, sizeof(qbuf))
+                || !EVP_DigestUpdate(ctx, d_intr, sizeof(d_intr)))
+            goto err;
+
+        if (odd) {
+            if (!EVP_DigestUpdate(ctx, path, n)
+                || !EVP_DigestUpdate(ctx, Tc, n))
+                goto err;
+        } else {
+            if (!EVP_DigestUpdate(ctx, Tc, n)
+                || !EVP_DigestUpdate(ctx, path, n))
+                goto err;
+        }
+        /*
+         * Tc = parent Hash, which is either the left or right child for the next
+         * level up (node_num determines if it is left or right).
+         */
+        if (!EVP_DigestFinal_ex(ctx, Tc, NULL))
+            goto err;
+        path += n;
+    }
+    ret = 1;
+err:
+    return ret;
+}
+
+/*
+ * @brief LMS signature verification.
+ * See RFC 8554 Section 5.4.2. Algorithm 6: Steps 3 & 4
+ *
+ * @param lms_sig Is a valid decoded LMS_SIG signature object.
+ * @param pub Is a valid LMS public key object.
+ * @param md Contains the fetched digest to be used for Hash operations
+ * @param msg A message to verify
+ * @param msglen The size of |msg|
+ * @returns 1 if the verification succeeded, or 0 otherwise.
+ */
+int ossl_lms_sig_verify(const LMS_SIG *lms_sig, const LMS_KEY *pub,
+                        const EVP_MD *md,
+                        const unsigned char *msg, size_t msglen)
+{
+    int ret = 0;
+    EVP_MD_CTX *ctx = NULL, *ctxIq = NULL;
+    EVP_MD_CTX *ctxI;
+    unsigned char Kc[LMS_MAX_DIGEST_SIZE];
+    unsigned char Tc[LMS_MAX_DIGEST_SIZE];
+    unsigned char qbuf[4];
+    unsigned char d_leaf[sizeof(uint16_t)];
+    const LMS_PARAMS *lms_params = pub->lms_params;
+    uint32_t n = lms_params->n;
+    uint32_t node_num;
+
+    ctx = EVP_MD_CTX_create();
+    ctxIq = EVP_MD_CTX_create();
+    if (ctx == NULL || ctxIq == NULL)
+        goto err;
+
+    if (!evp_md_ctx_init(ctxIq, md, lms_sig->params))
+        goto err;
+    /*
+     * Algorithm 6a: Step 3.
+     * Calculate a candidate public key |Kc| using the lmots_signature, message,
+     * and the identifiers I, q
+     */
+    if (!ossl_lm_ots_compute_pubkey(ctx, ctxIq, &lms_sig->sig,
+                                    pub->ots_params, pub->Id,
+                                    lms_sig->q, msg, msglen, Kc))
+        goto err;
+
+    /*
+     * Algorithm 6a: Step 4
+     * Compute the candidate LMS root value Tc
+     */
+    if (!ossl_assert(lms_sig->q < (uint32_t)(1 << lms_params->h)))
+        return 0;
+    node_num = (1 << lms_params->h) + lms_sig->q;
+
+    U32STR(qbuf, node_num);
+    U16STR(d_leaf, OSSL_LMS_D_LEAF);
+    ctxI = ctxIq;
+    /* Tc = H(I || u32str(node_num) || u16str(D_LEAF) || Kc) */
+    if (!EVP_DigestInit_ex2(ctx, NULL, NULL)
```

Comment:
So we are relying here on the side effect of `ossl_lm_ots_compute_pubkey` that it leaves `ctx` initialised with the `md`? Seems a bit unexpected.

Reply:
IMO ok but desires a comment.

Reply:
Comment added.